src/components/uv-scroll.js

```diff
@@ -1,20 +1,79 @@
+const textureToData = new Map();
+const registeredTextures = [];
+
+export class UVScrollSystem {
+  tick(_t, dt) {
+    for (let i = 0; i < registeredTextures.length; i++) {
+      const map = registeredTextures[i];
+      const { offset, instances } = textureToData.get(map);
+      const { component } = instances[0];
+
+      offset.addScaledVector(component.data.speed, dt / 1000);
+
+      const increment = component.data.increment;
+      map.offset.x = increment.x ? offset.x - (offset.x % increment.x) : offset.x;
+      map.offset.y = increment.y ? offset.y - (offset.y % increment.y) : offset.y;
+
+      // TODO this should be handled by MobileStandardMaterial itself
+      let matrixUpdated = false;
+      for (let j = 0; j < instances.length; j++) {
+        const mesh = instances[j].mesh;
+        if (mesh.material.isMobileStandardMaterial) {
+          if (!matrixUpdated) {
+            map.updateMatrix();
+            matrixUpdated = true;
+          }
+          mesh.material.uniforms.uvTransform.value.copy(map.matrix);
+        }
+      }
+    }
+  }
+}
+
 /**
  * Animate the UV offset of a mesh's material
  * @component uv-scroll
  */
 AFRAME.registerComponent("uv-scroll", {
   schema: {
-    speed: { type: "vec2", default: { x: 0, y: 0 } }
+    speed: { type: "vec2", default: { x: 0, y: 0 } },
+    increment: { type: "vec2", default: { x: 0, y: 0 } }
   },
-  tick(_t, dt) {
+  play() {
     const mesh = this.el.getObject3D("mesh") || this.el.getObject3D("skinnedmesh");
     const material = mesh && mesh.material;
-    if (!material || !material.map) return;
-    material.map.offset.addScaledVector(this.data.speed, dt / 1000);
+    if (material) {
+      const offset = new THREE.Vector2();
+      // We store mesh here instead of the material directly because we end up swapping out the material in injectCustomShaderChunks.
+      // We need material in the first place because of MobileStandardMaterial
+      const instance = { component: this, mesh };
+
+      this.instance = instance;
+      this.map = material.map;
+
+      if (!textureToData.has(material.map)) {
+        textureToData.set(material.map, {
+          offset,
+          instances: [instance]
+        });
+        registeredTextures.push(material.map);
+      } else {
+        console.warn(
+          "Multiple uv-scroll instances added to objects sharing a texture, only the speed/increment from the first one will have any effect"
+        );
+        textureToData.get(material.map).instances.push(instance);
+      }
+    }
+    // }, 0);
+  },
 
-    // TODO this should be handled by MobileStandardMaterial itself
-    if (material.isMobileStandardMaterial) {
-      material.refreshUniforms();
+  pause() {
+    const instances = textureToData.get(this.map).instances;
```

Comment:
Do we need a guard here like the `if (material)` guard in `play`? https://github.com/mozilla/hubs/pull/2597/files#diff-5aff2dd73415c1d2a28e0d360f143652R45